File: scripts/generate_catalog_index.py
#!/usr/bin/env python3
import csv, json, re
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
CSV_PATH = ROOT / 'data' / 'tcu_courses.csv'
if not CSV_PATH.exists():
    # fallback to old filename
    legacy = ROOT / 'data' / 'courses-report.2026-02-20.csv'
    if legacy.exists():
      CSV_PATH = legacy

OUT_PATH = ROOT / 'data' / 'catalog_index.json'

CREDIT_RE = re.compile(r'^\s*(\d+(?:\.\d+)?)\s*(?:hours?|credits?)\.?', re.IGNORECASE)
PREREQ_RE = re.compile(r'Prerequisites?:\s*(.*)', re.IGNORECASE | re.DOTALL)

catalog = {}

with CSV_PATH.open(newline='', encoding='utf-8-sig') as f:
    reader = csv.DictReader(f)
    for row in reader:
        subject = (row.get('Subject Code') or '').strip().upper()
        number = (row.get('Catalog Number') or '').strip()
        desc = (row.get('Description') or '').strip()
        if not subject or not number:
            continue
        code = f"{subject} {number}"

        credits = None
        m = CREDIT_RE.match(desc)
        if m:
            credits = float(m.group(1))
            if credits.is_integer():
                credits = int(credits)

        prereq_text = None
        pm = PREREQ_RE.search(desc)
        if pm:
            prereq_text = pm.group(1).strip()

        catalog[code] = {
            'subject': subject,
            'number': number,
            'description': desc,
            'credits': credits,
            'prereqText': prereq_text,
        }

OUT_PATH.write_text(json.dumps(catalog, indent=2, sort_keys=True), encoding='utf-8')
print(f'Wrote {len(catalog)} courses to {OUT_PATH}')

File: scripts/generate_catalog_index.js
#!/usr/bin/env node
const fs = require('fs');
const path = require('path');

const ROOT = path.join(__dirname, '..');
const DATA_DIR = path.join(ROOT, 'data');

function resolveInputCsv() {
  const preferred = path.join(DATA_DIR, 'tcu_courses.csv');
  if (fs.existsSync(preferred)) return preferred;

  const fallback = fs
    .readdirSync(DATA_DIR)
    .filter((name) => /^courses-report.*\.csv$/i.test(name))
    .sort()
    .pop();

  if (fallback) return path.join(DATA_DIR, fallback);
  throw new Error('No source CSV found. Expected data/tcu_courses.csv or data/courses-report*.csv');
}

function parseCsvLine(line) {
  const cells = [];
  let current = '';
  let inQuotes = false;

  for (let i = 0; i < line.length; i += 1) {
    const ch = line[i];
    if (ch === '"') {
      if (inQuotes && line[i + 1] === '"') {
        current += '"';
        i += 1;
      } else {
        inQuotes = !inQuotes;
      }
      continue;
    }
    if (ch === ',' && !inQuotes) {
      cells.push(current);
      current = '';
      continue;
    }
    current += ch;
  }

  cells.push(current);
  return cells;
}

function parseCsv(content) {
  const lines = content.replace(/^\uFEFF/, '').split(/\r?\n/).filter((line) => line.length > 0);
  if (lines.length === 0) return [];

  const headers = parseCsvLine(lines[0]);
  return lines.slice(1).map((line) => {
    const values = parseCsvLine(line);
    const row = {};
    headers.forEach((header, idx) => {
      row[header] = values[idx] || '';
    });
    return row;
  });
}

const CREDIT_RE = /^\s*(\d+(?:\.\d+)?)\s*(?:hours?|credits?)\.?/i;
const PREREQ_RE = /Prerequisites?:\s*(.*)/i;

function normalizeCredits(description) {
  const match = CREDIT_RE.exec(description || '');
  if (!match) return null;
  const value = Number(match[1]);
  if (!Number.isFinite(value)) return null;
  return Number.isInteger(value) ? value : value;
}

function buildCatalog(rows) {
  const catalog = {};
  rows.forEach((row) => {
    const subject = String(row['Subject Code'] || '').trim().toUpperCase();
    const number = String(row['Catalog Number'] || '').trim();
    const description = String(row.Description || '').trim();
    if (!subject || !number) return;

    const code = `${subject} ${number}`;
    const prereqMatch = PREREQ_RE.exec(description);

    catalog[code] = {
      subject,
      number,
      description,
      credits: normalizeCredits(description),
      prereqText: prereqMatch ? prereqMatch[1].trim() : null,
    };
  });
  return catalog;
}

function main() {
  const csvPath = resolveInputCsv();
  const outPath = path.join(DATA_DIR, 'catalog_index.json');

  const csvText = fs.readFileSync(csvPath, 'utf8');
  const rows = parseCsv(csvText);
  const catalog = buildCatalog(rows);

  const sortedCatalog = Object.fromEntries(Object.entries(catalog).sort(([a], [b]) => a.localeCompare(b)));
  fs.writeFileSync(outPath, `${JSON.stringify(sortedCatalog, null, 2)}\n`, 'utf8');
  console.log(`Wrote ${Object.keys(catalog).length} courses to ${outPath}`);
}

main();
